### PR TITLE
python3Packages.elegy: init at 0.8.4

### PR DESCRIPTION
**Pull request by ndl**
###### Motivation for this change

Finishes the "extra JAX libraries / frameworks" implementation, see the discussion in #152754 I also had to adjust `relax-deps.patch` for `treex` as due to PyYAML version update in `nixpkgs` it was out of allowed range.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/python-modules/elegy/default.nix
Comment (samuela) [inline pkgs/development/python-modules/elegy/default.nix:65]: potentially related: 
* https://github.com/tensorflow/tensorflow/issues/38205
* "Fix broken dependency (Optional)" section of https://nekodaemon.com/2021/08/04/Easy-way-to-debug-TensorFlow-XLA-Compiler-using-VSCode/

I believe it's coming from this bit here: https://cs.github.com/tensorflow/tensorflow/blob/10ac4ccb57af478fde1f6d479beb0dfa1aa47b8a/tensorflow/compiler/xla/service/compiler.cc?q=Could+not+find+compiler+for+platform+Host#L112-L115

**Review by samuela — APPROVED** [review]

**Comment by samuela** [comment]
It would be nice to track down the cause of the failing test, but I don't want that to be a blocker for merging this PR. So unless you'd like me to hold off @ndl, I'll merge in a day or two unless anyone else objects.

**Comment by ndl** [comment]
Yeah, agree it would be nice to track it down, but realistically I'm unlikely to look into this soon + it's not even clear if this would happen outside the tests, so I'd prefer to leave it "as is" for now.

**Comment by samuela** [comment]
Unfortunately I believe nixpkgs-review is stuck in some kind of live lock building one of the dependencies for this PR. It's been stuck on this step:
```
[1/19/23 built (7 failed), 381 copied (3912.8/3914.3 MiB), 1382.7 MiB DL] building python3.9-einops-0.3.2 (installCheckPhase): tests.test_layers.test_reduce_symbolic ... ok
```
for around half an hour now.

I know that this PR doesn't affect einops directly but we ought to get to the bottom of this nonetheless.

**Comment by samuela** [comment]
Created for the einops issue: https://github.com/NixOS/nixpkgs/issues/154251. Not sure that we need to worry about that for now though.

@ndl How did you manage to build elegy if the einops dep is failing?

**Comment by ndl** [comment]
`einops` test issue could be either non-deterministic or dependencies-dependent, hence I guess I got 'lucky' and it built locally the last time? I'm able to reproduce it locally at current `nixpkgs` master, though.

Logs show that `pytorch` is stuck at the inference step with multiple hundreds % of CPU usage, this seems similar to https://discuss.pytorch.org/t/pytorch-cpu-hangs-on-nn-linear/17748/4 and setting `OMP_NUM_THREADS=1`, as suggested in that thread, does seem to help. Added this work-around to this PR, please check if that helps on your side as well.

**Comment by samuela** [comment]
Result of `nixpkgs-review pr 154115` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages failed to build:</summary>
  <ul>
    <li>python310Packages.einops</li>
    <li>python310Packages.elegy</li>
    <li>python310Packages.treex</li>
  </ul>
</details>
<details>
  <summary>3 packages built:</summary>
  <ul>
    <li>python39Packages.einops</li>
    <li>python39Packages.elegy</li>
    <li>python39Packages.treex</li>
  </ul>
</details>

**Comment by samuela** [comment]
Nicely done! It looks like that fixed it for me. This PR may now supersede https://github.com/NixOS/nixpkgs/pull/152291. cc @veprbl @yl3dy 

Perhaps we should split this out into two PRs though? What does everyone think?

Python 3.10 is still busted but I suspect that's for some dumb reason out of scope here.

**Review by veprbl — APPROVED** [review]
LGTM

**Review by yl3dy — APPROVED** [review]
LGTM with respect to `einops`, thanks a lot!

**Comment by samuela** [comment]
ok as long as it looks good to @veprbl and @yl3dy, doing one PR is good with me

**Pull request merged**
